lib/tpm2_openssl.c

```diff
@@ -710,46 +696,80 @@ int tpm2_ossl_curve_to_nid(TPMI_ECC_CURVE curve) {
     return -1;
 }
 
-static bool load_public_ECC_from_key(EC_KEY *k, TPM2B_PUBLIC *pub) {
+static bool load_public_ECC_from_key(EVP_PKEY *key, TPM2B_PUBLIC *pub) {
 
+    BIGNUM *y = NULL;
+    BIGNUM *x = NULL;
+    int nid;
+    unsigned keysize;
     bool result = false;
 
-    BIGNUM *y = BN_new();
-    BIGNUM *x = BN_new();
-    if (!x || !y) {
-        LOG_ERR("oom");
-        goto out;
-    }
-
     /*
      * Set the algorithm type
      */
     pub->publicArea.type = TPM2_ALG_ECC;
+    TPMS_ECC_PARMS *pp = &pub->publicArea.parameters.eccDetail;
 
     /*
-     * Get the curve type
+     * Get the curve type and the public key (X and Y)
      */
-    const EC_GROUP *group = EC_KEY_get0_group(k);
-    int nid = EC_GROUP_get_curve_name(group);
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+    EC_KEY *k = EVP_PKEY_get0_EC_KEY(key);
+    if (!k) {
+        LOG_ERR("Could not retrieve ECC key");
+        goto out;
+    }
 
-    TPMS_ECC_PARMS *pp = &pub->publicArea.parameters.eccDetail;
-    TPM2_ECC_CURVE curve_id = ossl_nid_to_curve(nid); // Not sure what lines up with NIST 256...
-    if (curve_id == TPM2_ALG_ERROR) {
+    y = BN_new();
+    x = BN_new();
+    if (!x || !y) {
+        LOG_ERR("oom");
         goto out;
     }
 
-    pp->curveID = curve_id;
+    const EC_GROUP *group = EC_KEY_get0_group(k);
+    nid = EC_GROUP_get_curve_name(group);
+    keysize = (EC_GROUP_get_degree(group) + 7) / 8;
 
-    /*
-     * Set the unique data to the public key.
-     */
     const EC_POINT *point = EC_KEY_get0_public_key(k);
 
     int ret = EC_POINT_get_affine_coordinates_tss(group, point, x, y, NULL);
     if (!ret) {
         LOG_ERR("Could not get X and Y affine coordinates");
         goto out;
     }
+#else
+    char curveName[80];
+
+    int rc = EVP_PKEY_get_utf8_string_param(key, OSSL_PKEY_PARAM_GROUP_NAME,
```

Comment:
Should rc be unsigned?

Reply:
The documentation says the result is `int`. https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_get_utf8_string_param.html

Do you see a reason why `rc` should be unsigned?

Reply:
I see that makes sense. Let's stick with int.
I thought the return was never going to be negative.